src/test/java/com/betterlesson/learningwalk/services/LabServiceTests.java

```
@@ -0,0 +1,154 @@
+    InputStream targetStream = new FileInputStream(initialFile);
+    when(mockClientUtility.getHttpClient()).thenReturn(mockHttpClient);
+    when(mockHttpResponse.getStatusLine())
+        .thenReturn(new BasicStatusLine(HttpVersion.HTTP_1_1, 200, "Ok"));
+    when(mockEntity.getContent()).thenReturn(targetStream);
+    when(mockHttpResponse.getEntity()).thenReturn(mockEntity);
+    when(mockHttpClient.execute(any(HttpUriRequest.class))).thenReturn(mockHttpResponse);
+    when(mockObjectMapper.readValue(any(FileInputStream.class), eq(LabDataForLearningWalk.class)))
+        .thenThrow(IOException.class);
+
+    assertThrows(
+        LabDataSerializationException.class,
+        () -> {
+          mockLabService.getLabDataForLearningWalk(12L, 123L, "");
+        });
+  }
+
+  @Test
+  public void should_return_lab_data_if_the_lab_data_is_good() throws Exception {
+
+    File initialFile =
+        new File(
+            "src/test/java/com/betterlesson/learningwalk/testfiles/learningwalkdatafromlab.json");
+    InputStream targetStream = new FileInputStream(initialFile);
+    when(mockClientUtility.getHttpClient()).thenReturn(mockHttpClient);
+    when(mockHttpResponse.getStatusLine())
+        .thenReturn(new BasicStatusLine(HttpVersion.HTTP_1_1, 200, "Ok"));
+    when(mockEntity.getContent()).thenReturn(targetStream);
+    when(mockHttpResponse.getEntity()).thenReturn(mockEntity);
+    when(mockHttpClient.execute(any(HttpUriRequest.class))).thenReturn(mockHttpResponse);
+    // No point testing a third party serializer, so I am mocking out the response here
+    when(mockObjectMapper.readValue(any(FileInputStream.class), eq(LabDataForLearningWalk.class)))
+        .thenReturn(getMockLabDataForLearningWalk());
+    try {
+      mockLabService.getLabDataForLearningWalk(123L, 123L, "");
+    } catch (LabUnavailableException | LabDataSerializationException e) {
+      fail("Should not have thrown any exception");
+    }
```

nit: why try/catch here? Could we not just let the exception be thrown?

---

Good point. The test should fail if the code throws an exception at this point.